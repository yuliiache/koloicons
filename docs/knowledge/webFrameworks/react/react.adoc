== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







