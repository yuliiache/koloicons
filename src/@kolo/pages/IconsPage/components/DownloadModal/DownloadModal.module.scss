@import "@kolo/design";

.modalContainer {
  color: $secFontColor;
}
.modalHeader {
  margin: 0 0 24px 0;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  letter-spacing: normal;
  color: $secFontColor;

  h3 {
    font-size: 16px;
  }
}

.modalMainContent {
  display: flex;
  flex-direction: column;
}
.modalBody {
  width: 100%;
  gap: 16px;
  display: flex;
}
