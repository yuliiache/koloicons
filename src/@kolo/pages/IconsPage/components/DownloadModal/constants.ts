export const ICON_KEY = 'icon';
