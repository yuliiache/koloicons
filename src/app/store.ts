import {getReducers} from '@kolo/core/reducers/getReducers';
import {rootSagas} from '@kolo/core/sagas/rootSagas';
import {configureStore} from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';

const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
  reducer: getReducers(),
  devTools: process.env.NODE_ENV !== 'production',
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(sagaMiddleware),
});

sagaMiddleware.run(rootSagas);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
