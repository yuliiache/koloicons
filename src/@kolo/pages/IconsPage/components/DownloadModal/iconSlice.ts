import {Icon} from '@kolo/facade/getIcon/types';
import {createSlice, PayloadAction} from '@reduxjs/toolkit';

import {ICON_KEY} from './constants';

const initialState: {
  iconId: string;
  icon: Icon;
} = {
  iconId: '',
  icon: {name: '', category: {name: '', number: 0, id: ''}, size: '', style: '', types: [], id: '', tags: []},
};

const iconSlice = createSlice({
  name: ICON_KEY,
  initialState,
  reducers: {
    getIconStart: (state, action: PayloadAction<string>) => ({
      ...state,
      iconId: action.payload,
    }),
    getIconSuccess: (state, action: PayloadAction<Icon>) => ({
      ...state,
      icon: action.payload,
    }),
    getIconFail: () => ({
      icon: initialState.icon,
      iconId: initialState.iconId,
    }),
  },
});

export const {getIconStart, getIconSuccess, getIconFail} = iconSlice.actions;

export default iconSlice.reducer;
