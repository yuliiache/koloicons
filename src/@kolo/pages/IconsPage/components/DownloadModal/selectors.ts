import {Icon} from '@kolo/facade/getIcon/types';
import {RootState} from 'app/store';
import {createSelector, Selector} from 'reselect';

import {ICON_KEY} from './constants';

type IconSelectorState = Selector<RootState, Icon>;

const getIcon = (state: RootState) => state[ICON_KEY];

export const IconSelector: IconSelectorState = createSelector(getIcon, (selectedIcon) => selectedIcon.icon);
