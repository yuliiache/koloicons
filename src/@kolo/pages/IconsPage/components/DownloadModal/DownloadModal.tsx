import {isAuthenticatedSelector, userSelector} from '@kolo/pages/LoginPage/selectors';
import {useAppSelector} from 'app/hooks';
import {FC} from 'react';

import DownloadOptions from '../DownloadOptions/DownloadOptions';
import DownloadModalIcon from './components/DownloadModalIcon/DownloadModalIcon';
import styles from './DownloadModal.module.scss';
import {ModalIconProps} from './types/DownloadModalProps';

const DownloadModal: FC<ModalIconProps> = ({iconId, isFree}) => {
  const {isPro} = useAppSelector(userSelector);
  const isAuthenticated = useAppSelector(isAuthenticatedSelector);

  return (
    <div className={styles.modalContainer}>
      <div className={styles.modalMainContent}>
        <div className={styles.modalHeader}>
          <h3>Download Icon</h3>
        </div>
        <div className={styles.modalBody}>
          <DownloadModalIcon
            iconId={iconId}
            isFree={isFree}
          />
          <DownloadOptions
            isAuthenticated={isAuthenticated}
            isPro={isPro}
          />
        </div>
      </div>
    </div>
  );
};

export default DownloadModal;
