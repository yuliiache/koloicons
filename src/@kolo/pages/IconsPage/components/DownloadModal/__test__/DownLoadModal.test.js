import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';
import {render} from '@testing-library/react';
import store from 'app/store';

import DownloadModal from '../DownloadModal';

describe('Download Modal', () => {
  it('renders correctly and matches snapshot', () => {
    const {asFragment} = render(
      <Provider store={store}>
        <BrowserRouter>
          <DownloadModal />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
