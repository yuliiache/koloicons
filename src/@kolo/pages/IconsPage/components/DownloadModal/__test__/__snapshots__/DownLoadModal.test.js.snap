// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download Modal renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="modalContainer"
  >
    <div
      class="modalMainContent"
    >
      <div
        class="modalHeader"
      >
        <h3>
          Download Icon
        </h3>
      </div>
      <div
        class="modalBody"
      >
        <div
          class="modalIcon"
        >
          <div
            class="modalIconHead"
          >
            <h2
              class="modalHeadTitle"
            />
            <div
              class="modalColorPicker"
            >
              <label
                class="label big"
              >
                <div
                  class="colorPointer"
                  style="background-color: rgb(0, 0, 0);"
                />
                <input
                  class="colorInput"
                  data-testid="colorPickerComponent"
                  id="colorPicker"
                  type="color"
                  value="#000000"
                />
              </label>
            </div>
          </div>
          <ul
            class="iconTags"
          />
          <div
            class="modalIconImage"
          >
            <div
              class="adaptiveImage"
            >
              <div
                class="adaptiveImageLoader"
              >
                <svg
                  data-testid="icon-loader"
                  height="24"
                  version="1.1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                >
                  <g
                    fill="none"
                    fill-rule="evenodd"
                    stroke="none"
                    stroke-width="1"
                  >
                    <g
                      fill="#DFE5F0"
                      transform="translate(-12.000000, -12.000000)"
                    >
                      <g
                        transform="translate(24.000000, 24.000000) rotate(360.000000) translate(-24.000000, -24.000000) translate(12.000000, 12.000000)"
                      >
                        <path
                          d="M9.28068148,14.9335046 L9.18812903,15.1332387 C8.34851613,17.0413548 8.39903226,19.3093548 9.52258065,21.2516129 C10.2548787,22.5173105 11.1845921,23.3960079 12.1811402,23.9992779 L12,24 C9.87718736,24 7.88311694,23.448789 6.15310874,22.481687 C5.59137724,21.4607284 5.22580645,20.2349738 5.22580645,18.7741935 C5.22580645,16.4516297 6.39561372,14.4014611 8.17797777,13.1809384 C8.36083959,13.7755013 8.67886092,14.3105373 9.09596264,14.7507503 L9.28068148,14.9335046 Z M18.7741935,18.7741935 C20.2349481,18.7741935 21.4612813,18.4088045 22.4825866,17.8475631 C21.454853,19.6837516 19.9591672,21.2229234 18.156604,22.3024323 C16.9916263,22.2779506 15.7474587,21.9813011 14.4822581,21.2520968 C12.4687963,20.0918018 11.2784223,18.0544797 11.1128317,15.9004291 C11.3976552,15.9658272 11.6948369,16 12,16 C12.4112778,16 12.8080583,15.9379293 13.1814846,15.8226448 C14.4014611,17.6041997 16.4516297,18.7741935 18.7741935,18.7741935 Z M9.06848284,9.27957578 L8.98113257,9.37570619 C8.55616581,9.86415284 8.24779071,10.4567243 8.09935394,11.1100738 C5.94682452,11.2762301 3.90820183,12.4672741 2.7483871,14.4774194 C2.017736,15.7442666 1.72130398,16.9903421 1.69700407,18.1568826 C0.686951133,16.4691533 0.0792399695,14.5127291 0.00722159332,12.420267 L0.003,12.167 L0.145305469,11.943179 C0.742882636,11.0336466 1.58175506,10.1903812 2.7483871,9.51774194 C4.75883045,8.35775519 7.12015471,8.34559692 9.06848284,9.27957578 Z M22.3075574,5.85065135 C23.3148491,7.53647076 23.9209672,9.49052592 23.9927973,11.5802793 L23.997,11.836 L23.854932,12.060861 C23.2577031,12.9666264 22.4201746,13.8063674 21.2564516,14.4774194 C19.242189,15.6405809 16.8804974,15.6534911 14.9315387,14.7188987 C15.4028001,14.2136397 15.7433172,13.5849087 15.9013262,12.8869273 C18.0545037,12.7215143 20.0932805,11.5312001 21.2564516,9.51774194 C21.9844337,8.25529197 22.2813633,7.01357201 22.3075574,5.85065135 Z M12,0 C14.1232205,0 16.1176493,0.551422815 17.8478884,1.51887044 C18.4087128,2.53922155 18.7741935,3.76497495 18.7741935,5.22580645 C18.7741935,7.54851939 17.6057998,9.59910483 15.8231088,10.8196409 C15.6131743,10.1400048 15.2280213,9.53859655 14.7203146,9.06741951 C15.6544112,7.11966111 15.6426382,4.75921169 14.4822581,2.7483871 C13.7510281,1.48430505 12.8218587,0.605105262 11.8258083,0.000698636871 L12,0 Z M5.84641636,1.69575951 C7.01154896,1.71928687 8.25658637,2.01605099 9.52258065,2.7483871 C11.5333487,3.90826062 12.7245078,5.94599317 12.8896301,8.10046875 C12.6037318,8.03433244 12.3058753,8 12,8 C11.589092,8 11.1926553,8.06195908 10.8195229,8.17704434 C9.59907466,6.39418303 7.5485023,5.22580645 5.22580645,5.22580645 C3.76440244,5.22580645 2.53861964,5.59141038 1.51760638,6.1524873 C2.54626189,4.31466557 4.04282978,2.77516228 5.84641636,1.69575951 Z"
                        />
                      </g>
                    </g>
                  </g>
                </svg>
              </div>
              <picture
                class="adaptiveImagePicture"
              >
                <img
                  alt=""
                  data-isfree="notfree"
                  data-testid=""
                  height="240"
                  loading="lazy"
                  src=""
                  srcset=""
                  width="240"
                />
              </picture>
            </div>
          </div>
          <div
            class="modalIconFooter"
          >
            <div
              class="modalIconFooterCategory"
            >
              <svg
                data-testid="icon-folder"
                height="26"
                viewBox="0 0 26 26"
                width="26"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill-rule="evenodd"
                >
                  <path
                    d="M22,19 C22,20.1045695 21.1045695,21 20,21 L4,21 C2.8954305,21 2,20.1045695 2,19 L2,5 C2,3.8954305 2.8954305,3 4,3 L9,3 L11,6 L20,6 C21.1045695,6 22,6.8954305 22,8 L22,19 Z"
                    fill="#e9edf7"
                  />
                  <path
                    d="M11.5351838,5 L20,5 C21.6568542,5 23,6.34314575 23,8 L23,19 C23,20.6568542 21.6568542,22 20,22 L4,22 C2.34314575,22 1,20.6568542 1,19 L1,5 C1,3.34314575 2.34314575,2 4,2 L9,2 C9.33435319,2 9.64658452,2.16710114 9.83205029,2.4452998 L11.5351838,5 Z M8.46481624,4 L4,4 C3.44771525,4 3,4.44771525 3,5 L3,19 C3,19.5522847 3.44771525,20 4,20 L20,20 C20.5522847,20 21,19.5522847 21,19 L21,8 C21,7.44771525 20.5522847,7 20,7 L11,7 C10.6656468,7 10.3534155,6.83289886 10.1679497,6.5547002 L8.46481624,4 Z"
                    fill="#717886"
                  />
                </g>
              </svg>
              <h3
                class="modalIconFooterCategoryName"
              />
            </div>
            <button
              class="button button-full-width button-basic button-center"
              data-testid="buttonComponent"
              type="button"
            >
              <span
                data-testid="buttonComponentText"
              >
                Show icons from this category
              </span>
            </button>
          </div>
        </div>
        <div
          class="DownloadPage"
        >
          <div
            class="formats__wrapper"
          >
            <div
              class="formats__text--desktop"
            >
              <div
                class="formats__text__unlock-icon"
              >
                <svg
                  data-testid="icon-unlock"
                  height="50"
                  viewBox="0 0 48 48"
                  width="50"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    fill-rule="evenodd"
                  >
                    <circle
                      cx="24"
                      cy="24"
                      fill="#FFF"
                      r="24"
                    />
                    <g
                      fill-rule="nonzero"
                      transform="translate(14 13)"
                    >
                      <path
                        d="M15.9796352,4.80402869 C16.0904588,5.34500402 15.7378583,5.87260082 15.1920799,5.98244911 C14.6463016,6.09229741 14.1140206,5.74279989 14.003197,5.20182456 C13.5908301,3.18889267 11.7067105,1.8131833 9.64495448,2.01960876 C7.6519238,2.21915334 6.11980921,3.83434832 6.02176252,5.79809324 L6.01677524,6.00243885 L6.016,9 L17,9 C18.5976809,9 19.9036609,10.24892 19.9949073,11.8237272 L20,12 L20,19 C20,20.6568542 18.6568542,22 17,22 L17,22 L3,22 C1.34314575,22 0,20.6568542 0,19 L0,19 L0,12 C0,10.3431458 1.34314575,9 3,9 L3,9 L3.999,9 L4,6.00344236 C3.9968823,2.92242739 6.34963804,0.340343094 9.44227144,0.0307049798 C12.5349055,-0.278933203 15.3610849,1.78463085 15.9796352,4.80402869 Z M17,11 L3,11 C2.44771525,11 2,11.4477153 2,12 L2,12 L2,19 C2,19.5522847 2.44771525,20 3,20 L3,20 L17,20 C17.5522847,20 18,19.5522847 18,19 L18,19 L18,12 C18,11.4477153 17.5522847,11 17,11 L17,11 Z"
                        fill="#008f4a"
                      />
                      <path
                        d="M3,11 L17,11 C17.5522847,11 18,11.4477153 18,12 L18,19 C18,19.5522847 17.5522847,20 17,20 L3,20 C2.44771525,20 2,19.5522847 2,19 L2,12 C2,11.4477153 2.44771525,11 3,11 Z"
                        fill="#d8f0e6"
                      />
                    </g>
                  </g>
                </svg>
              </div>
              <h3
                class="formats__text__heading"
              >
                Unlock all Formats
              </h3>
              <p
                class="formats__text__details"
              >
                All icons in any size and format:
                <br />
                SVG and PNG
              </p>
              <div
                class="formats__text__benefits"
              >
                <a
                  class="link link--accent"
                  data-testid="internalLinkComponent"
                  href="/pricing"
                >
                  All Benefits
                </a>
              </div>
            </div>
            <a
              data-testid="buttonLinkComponent"
              href="/pricing"
            >
              <button
                class="button button-full-width button-accent button-center button-children"
                data-testid="buttonComponent"
                type="button"
              >
                <span
                  data-testid="buttonComponentText"
                >
                  Purchase
                </span>
                <span>
                  $2.50
                  <span
                    class="purchase-button__period"
                  >
                    /mo
                  </span>
                </span>
              </button>
            </a>
          </div>
          <div>
            <button
              class="button button-full-width button-primary button-left button-disabled button-children"
              disabled=""
              type="button"
            >
              <div
                class="buttonContent buttonContent-disabled"
              >
                <span>
                  <svg
                    data-testid="icon-download"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12.5,14 C13.0522847,14 13.5,14.4477153 13.5,15 C13.5,15.5128358 13.1139598,15.9355072 12.6166211,15.9932723 L12.5,16 L3.5,16 C2.94771525,16 2.5,15.5522847 2.5,15 C2.5,14.4871642 2.88604019,14.0644928 3.38337887,14.0067277 L3.5,14 L12.5,14 Z M7.4097362,12.4786616 L1.39390847,6.46323438 C1.05917337,6.12849928 1.05917337,5.58578644 1.39390847,5.25105133 C1.70074899,4.94421082 2.18234857,4.91864077 2.5183156,5.1743412 L2.60609153,5.25105133 L7.14185714,9.786 L7.14285714,0.857142857 C7.14285714,0.420170308 7.46984444,0.0595697346 7.89248182,0.00667835971 L8,-5.5067062e-14 C8.43697255,-5.5067062e-14 8.79757312,0.3269873 8.8504645,0.749624678 L8.85714286,0.857142857 L8.85685714,9.787 L13.3939085,5.25105133 C13.700749,4.94421082 14.1823486,4.91864077 14.5183156,5.1743412 L14.6060915,5.25105133 C14.912932,5.55789185 14.9385021,6.03949143 14.6828017,6.37545845 L14.6060915,6.46323438 L8.60609153,12.4632344 C8.59911788,12.470208 8.59205396,12.4770364 8.58490354,12.4837195 C8.57154085,12.4961969 8.55759753,12.508409 8.54326621,12.5201656 L8.50375905,12.5507817 L8.50375905,12.5507817 C8.48730436,12.5626896 8.46997404,12.5743751 8.45220971,12.5854288 C8.43933536,12.5935202 8.42652778,12.6010216 8.41357462,12.6081627 L8.35712029,12.6365758 L8.35712029,12.6365758 C8.34303676,12.6430511 8.32891323,12.6490588 8.31466974,12.6546704 C8.25947488,12.676496 8.2012221,12.6927888 8.14083979,12.7027705 C8.1302706,12.7044516 8.11986535,12.7059776 8.10944028,12.7073121 L8,12.7142857 L7.89055972,12.7073121 L7.85276784,12.7016888 C7.82945578,12.6976532 7.80646644,12.6926758 7.78384042,12.6867974 C7.77506332,12.6844536 7.76561855,12.6818373 7.75621211,12.6790563 C7.73153679,12.6717884 7.70804836,12.6636629 7.68504738,12.6545709 C7.66767068,12.647713 7.65019549,12.6401466 7.63293694,12.6319712 C7.61798022,12.624824 7.60378498,12.6175788 7.5898251,12.6099565 C7.57431494,12.6015822 7.55851105,12.5923088 7.54295432,12.5824865 L7.49583076,12.5504036 L7.49583076,12.5504036 L7.45685714,12.52 L7.44371128,12.5092982 L7.44371128,12.5092982 L7.43885714,12.505 L7.45685714,12.52 L7.40485714,12.473 L7.43385714,12.501 L7.4097362,12.4786616 L7.4097362,12.4786616 Z"
                    />
                  </svg>
                </span>
                <span>
                  PNG
                </span>
              </div>
              <span
                class="additionalText"
              >
                60px
              </span>
            </button>
          </div>
          <div>
            <button
              class="button button-full-width button-primary button-left button-disabled"
              disabled=""
              type="button"
            >
              <div
                class="buttonContent buttonContent-disabled"
              >
                <span>
                  <svg
                    data-testid="icon-download"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12.5,14 C13.0522847,14 13.5,14.4477153 13.5,15 C13.5,15.5128358 13.1139598,15.9355072 12.6166211,15.9932723 L12.5,16 L3.5,16 C2.94771525,16 2.5,15.5522847 2.5,15 C2.5,14.4871642 2.88604019,14.0644928 3.38337887,14.0067277 L3.5,14 L12.5,14 Z M7.4097362,12.4786616 L1.39390847,6.46323438 C1.05917337,6.12849928 1.05917337,5.58578644 1.39390847,5.25105133 C1.70074899,4.94421082 2.18234857,4.91864077 2.5183156,5.1743412 L2.60609153,5.25105133 L7.14185714,9.786 L7.14285714,0.857142857 C7.14285714,0.420170308 7.46984444,0.0595697346 7.89248182,0.00667835971 L8,-5.5067062e-14 C8.43697255,-5.5067062e-14 8.79757312,0.3269873 8.8504645,0.749624678 L8.85714286,0.857142857 L8.85685714,9.787 L13.3939085,5.25105133 C13.700749,4.94421082 14.1823486,4.91864077 14.5183156,5.1743412 L14.6060915,5.25105133 C14.912932,5.55789185 14.9385021,6.03949143 14.6828017,6.37545845 L14.6060915,6.46323438 L8.60609153,12.4632344 C8.59911788,12.470208 8.59205396,12.4770364 8.58490354,12.4837195 C8.57154085,12.4961969 8.55759753,12.508409 8.54326621,12.5201656 L8.50375905,12.5507817 L8.50375905,12.5507817 C8.48730436,12.5626896 8.46997404,12.5743751 8.45220971,12.5854288 C8.43933536,12.5935202 8.42652778,12.6010216 8.41357462,12.6081627 L8.35712029,12.6365758 L8.35712029,12.6365758 C8.34303676,12.6430511 8.32891323,12.6490588 8.31466974,12.6546704 C8.25947488,12.676496 8.2012221,12.6927888 8.14083979,12.7027705 C8.1302706,12.7044516 8.11986535,12.7059776 8.10944028,12.7073121 L8,12.7142857 L7.89055972,12.7073121 L7.85276784,12.7016888 C7.82945578,12.6976532 7.80646644,12.6926758 7.78384042,12.6867974 C7.77506332,12.6844536 7.76561855,12.6818373 7.75621211,12.6790563 C7.73153679,12.6717884 7.70804836,12.6636629 7.68504738,12.6545709 C7.66767068,12.647713 7.65019549,12.6401466 7.63293694,12.6319712 C7.61798022,12.624824 7.60378498,12.6175788 7.5898251,12.6099565 C7.57431494,12.6015822 7.55851105,12.5923088 7.54295432,12.5824865 L7.49583076,12.5504036 L7.49583076,12.5504036 L7.45685714,12.52 L7.44371128,12.5092982 L7.44371128,12.5092982 L7.43885714,12.505 L7.45685714,12.52 L7.40485714,12.473 L7.43385714,12.501 L7.4097362,12.4786616 L7.4097362,12.4786616 Z"
                    />
                  </svg>
                </span>
                <span>
                  SVG
                </span>
              </div>
            </button>
          </div>
          <div
            class="signInMessage"
          >
            <p
              class="signInText"
            >
              Sign In to work with
              <span
                class="signInTitle"
                data-testid="signInMessageComponent"
              >
                 Sign In
              </span>
            </p>
            <a
              data-testid="buttonLinkComponent"
              href="/login"
            >
              <button
                class="button button-full-width button-basic button-center"
                data-testid="buttonComponent"
                type="button"
              >
                <span
                  data-testid="buttonComponentText"
                >
                  Sign In
                </span>
              </button>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
